authentication/oidc.go

```diff
@@ -45,71 +47,112 @@ type OIDCConfig struct {
 // Middleware is a convenience type for functions that wrap http.Handlers.
 type Middleware func(http.Handler) http.Handler
 
-// NewOIDC creates a single http.Handler and a set of Middlewares for all
+// OIDCHandlers holds handlers and a set of middlewares for all
 // tenants that is able to authenticate requests and provide the
 // authorization code grant flow for users.
-func NewOIDC(
-	logger log.Logger,
-	reg prometheus.Registerer,
-	prefix string,
-	configs []TenantOIDCConfig,
-) (http.Handler, map[string]Middleware, []error) {
-	handlers := map[string]http.Handler{}
-	middlewares := map[string]Middleware{}
-	warnings := make([]error, 0, len(configs))
+type OIDCHandlers struct {
+	mtx         sync.RWMutex
+	handlers    map[string]http.Handler
+	middlewares map[string]Middleware
 
+	logger          log.Logger
+	retryCountIncFn func()
+}
+
+// NewOIDCHandlers instantiates OIDC handlers.
+func NewOIDCHandlers(l log.Logger, reg prometheus.Registerer) *OIDCHandlers {
 	tenantErrors := prometheus.NewGauge(prometheus.GaugeOpts{
 		Namespace: "observatorium",
 		Subsystem: "api",
 		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Help:      "The number of attempts for which tenant OIDC provider instantiation failed.",
 	})
```

Comment:
With the retry logic, I think this metric is not very usable. WDYT about replacing it with a gauge that shows the current number of tenants undergoing retries (instead of the total number of attempts)? This way, we can alert if there is a tenant failing to instantiate for a long time, etc.

Reply:
Makes much more sense :+1: Will adjust.